src/sgame/sg_bot.cpp

```diff
@@ -263,6 +262,188 @@ bool G_BotSetDefaults( int clientNum, team_t team, int skill, Str::StringRef beh
 	return true;
 }
 
+struct botSkillTreeElement_t {
+	Str::StringRef name;
+	enum bot_skill enable;
+	int cost;
+	// function that determines if this skill is available
+	std::function<bool(const gentity_t *self, skillSet_t existing_skills)> predicate;
+	// skills that are unlocked once you unlock this one
+	std::vector<botSkillTreeElement_t> unlocked_skills;
```

Comment:
Could skill dependencies be moved into `predicate` and `unlocked_skills` removed? You already call the predicate on every skill in the list every time you choose a new skill anyway

Reply:
I guess? Initially I made it that way because dependencies are declared as a subelement like this which is IMO more readable:
```C++
	// humans
	{ "buy-armor",          BOT_H_BUY_ARMOR,   10, pred_human, {
		{ "prefer-armor", BOT_H_PREFER_ARMOR, 5, pred_human, {} },
	}},
```

And because otherwise this forces to write a lot of functions, or to do trickeries such as writing a `pred_depends_on()` and a `pred_and()` function to do the logical and so you can start doing weird things such as `std::function predicate = pred_and(pred_human, std::bind(pred_depends_on, BOT_H_BUY_ARMOR))`.

I also fear that if dependencies aren't displayed in a tree like that (allowing for some extra dependency if need be), the tree will become much harder to maintain because it wouldn't be possible to have a global view of what depends on what, unless someone draws it on paper when checking or unless we maintain a separate document.

Reply:
I guess I would do like `[](gentity_t* ent) { return G_Team(ent) == TEAM_HUMANS && ent->botMind->skills[BOT_H_PREFER_ARMOR]; }` instead of trying to compose predicates.

Heh, I didn't notice until now that the skills are literally stored in a tree. I don't see why it should be expected to be a tree and not a more general digraph, like maybe you have a skill that depends on both fighting and moving.

Anyway I'm not insisting on a particular change right here, just discussion.